nomad/heartbeat.go

```diff
@@ -173,6 +182,23 @@ func (h *nodeHeartbeater) invalidateHeartbeat(id string) {
 	}
 }
 
+// nodeStatusIsNotDown will query the Client for its current status, returning
+// true if we can reach the client and its status is not already down.
+//
+// This is only used in the context of a last attempt at making sure a client
+// is actually down - (a workaround for racy / misguided heartbeat timer).
+func (h *nodeHeartbeater) nodeStatusIsNotDown(id string) bool {
+	args := structs.NodeSpecificRequest{NodeID: id}
+	var reply structs.SingleNodeResponse
+	if err := findNodeConnAndForward(h.Server, id, "ClientStats.Stats", &args, &reply); err != nil {
+		// If we cannot reach the client then yeah, it's probably down for real.
+		h.logger.Info("unable to retrieve client status directly", "error", err)
+		return false
+	}
+	// Any status other than down is good enough to cancel the missed-heartbeat.
+	return reply.Node.Status != structs.NodeStatusDown
```

Comment:
If we cancel the missed heartbeat on `structs.NodeStatusDisconnected` will that cause disconnected clients to rejoin accidentally?